napari/layers/labels/labels.py

```diff
@@ -395,30 +402,43 @@ def data(self, data):
         self.events.data(value=self.data)
         self._set_editable()
 
+    @property
+    def features(self):
+        """DataFrame-like features table.
+
+        Treat this as a very simple pandas DataFrame that maps a feature name
+        to a 1D array-like. Do not rely on other parts of the pandas DataFrame API,
+        which may not be supported by future versions of napari.
+        """
+        return self._features
+
+    @features.setter
+    def features(
+        self,
+        features: Union[Dict[str, np.ndarray], pd.DataFrame],
+    ) -> None:
+        self._features = validate_features(features)
+        self._label_index = self._make_label_index(self._features)
+
     @property
     def properties(self) -> Dict[str, np.ndarray]:
         """dict {str: array (N,)}, DataFrame: Properties for each label."""
-        return self._properties
+        return features_to_properties(self._features)
 
     @properties.setter
     def properties(self, properties: Dict[str, Array]):
-        self._properties, self._label_index = self._prepare_properties(
-            properties
-        )
+        self._features = features_from_properties(properties=properties)
+        self._label_index = self._make_label_index(self._features)
         self.events.properties()
 
     @classmethod
-    def _prepare_properties(
-        cls, properties: Optional[Dict[str, Array]]
-    ) -> Tuple[Dict[str, np.ndarray], Dict[int, int]]:
-        properties = validate_properties(properties)
+    def _make_label_index(cls, features: pd.DataFrame) -> Dict[int, int]:
         label_index = {}
-        if 'index' in properties:
-            label_index = {i: k for k, i in enumerate(properties['index'])}
-        elif len(properties) > 0:
-            max_len = max(len(x) for x in properties.values())
-            label_index = {i: i for i in range(max_len)}
-        return properties, label_index
+        if 'index' in features:
+            label_index = {i: k for k, i in enumerate(features['index'])}
+        elif features.shape[1] > 0:
+            label_index = {i: i for i in range(features.shape[0])}
+        return label_index
```

Comment:
This is a really useful comment, as I mostly refactored this blind without thinking too much and trying to be safe.

I like the idea of requiring the user to set `DataFrame.index`, but my biggest concern would be that it may not be compatible with the initial version of the DataFrame protocol, as there is [a documented design decision to not support row names initially](https://data-apis.org/dataframe-protocol/latest/API.html#design-decisions), instead preferring an approaching like this where the index/label/name is a column.

But at least we could drop the implementation detail of `Labels._label_index` and use `DataFrame.index` while we are using pandas DataFrame as our current implementation.

Reply:
We decided to leave this alone for now, but will probably follow-up with a small change here.